sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/MessageBody.cs

```diff
@@ -162,5 +170,82 @@ private void Append(Data segment)
                 _segments.Add(memory.Slice(0, dataToAppend.Length));
             }
         }
+
+        /// <summary>
+        /// Eagerly copies the provided data segments into a single continuous buffer while still keeping around a list of the individual copied segments.
+        /// Important for the receive path in order to make sure the buffers managed by the underlying AMQP library can be released on dispose.
+        /// </summary>
+        private sealed class EagerCopyingMessageBodyWithManagedBuffer : MessageBody
+        {
+            private IList<ReadOnlyMemory<byte>> _segments;
+            private byte[] _continuousBuffer;
+            private ReadOnlyMemory<byte>? _writtenMemory;
+
+            internal EagerCopyingMessageBodyWithManagedBuffer(IEnumerable<Data> dataSegments)
+            {
+                Append(dataSegments);
+            }
+
+            protected override ReadOnlyMemory<byte> WrittenMemory => _writtenMemory ?? ReadOnlyMemory<byte>.Empty;
+
+            public override IEnumerator<ReadOnlyMemory<byte>> GetEnumerator()
+            {
+                return _segments.GetEnumerator();
+            }
+
+            private void Append(IEnumerable<Data> dataSegments)
+            {
+                int length = 0;
+                List<ReadOnlyMemory<byte>> convertedSegments = null;
+                foreach (var dataSegment in dataSegments)
+                {
+                    convertedSegments ??= new List<ReadOnlyMemory<byte>>();
+                    ReadOnlyMemory<byte> readOnlyMemory = Convert(dataSegment);
+                    convertedSegments.Add(readOnlyMemory);
+                    length += readOnlyMemory.Length;
+                }
+
+                if (convertedSegments == null)
+                {
+                    return;
+                }
+
+                _segments ??= new List<ReadOnlyMemory<byte>>();
+                _continuousBuffer = ArrayPool<byte>.Shared.Rent(length);
```

Comment:
Conceptually, I like the concept of renting the shared buffer over rolling our own.